### DISCUSSION TITLE
mixing 'flutter riverpod' with 'dart riverpod'  (mixing ref and ProviderContainer)

### DISCUSSION THREAD

**Discussion by 3SCadmin** [rrousselGit/riverpod#1136] [Q&A]
Hi,

I am just getting started with riverpod

I have a StateNotifierProvider. 
I want to interact with my StateNotifierProvider from "non-gui" parts of the app (`WidgetsBinding.instance!.addPostFrameCallback((_) {}`)

In your example code I discovered this: `final providerContainer = ProviderContainer();`

Using the above I was able to "trigger" a function inside my StateNotifierProvider without a Consumer widget.
BUT... it appears that using both ProviderScope and ProviderContainer();  creates 2 separate instances of my provider.

Best to show the full example code, to give this some reference:

```
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

void main() {
  runApp(ProviderScope(child: MartinApp()));
}

final indexProvider = StateNotifierProvider((ref) => displayOrderController(1));

class displayOrderController extends StateNotifier<int> {
  displayOrderController(int state) : super(state);

  add(int i) {
    state = state + i;
    print('new state= $state');
  }
}

final providerContainer = ProviderContainer();

class MartinApp extends StatefulWidget {
  const MartinApp({Key? key}) : super(key: key);

  @override
  _MartinAppState createState() => _MartinAppState();
}

class _MartinAppState extends State<MartinApp> {
  void initState() {
    print('initState called');
    super.initState();
    WidgetsBinding.instance!.addPostFrameCallback((_) {
      print("Loading screen completed");
      providerContainer.read(indexProvider.notifier).add(100);
    });
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: SafeArea(
          child: Center(
            child: Column(
              mainAxisAlignment: MainAxisAlignment.center,
              children: [
                Consumer(
                  builder: (context, ref, child) {
                    int num = ref.watch(indexProvider) as int;
                    return Text(num.toString());
                  },
                ),
                Consumer(builder: (context, ref, child) {
                  return ElevatedButton(
                      onPressed: () {
                        ref.read(indexProvider.notifier).add(5);
                      },
                      child: Text('Add 5 (ref)'));
                }),
                Consumer(builder: (context, ref, child) {
                  return ElevatedButton(
                      onPressed: () {
                        providerContainer.read(indexProvider.notifier).add(5);
                      },
                      child: Text('Add 5 (providerContainer)'));
                }),
              ],
            ),
          ),
        ),
      ),
    );
  }
}

```
If I alternately press each of my buttons 3 times a get the following debug lines:
```
I/flutter ( 6145): new state= 6
I/flutter ( 6145): new state= 106
I/flutter ( 6145): new state= 11
I/flutter ( 6145): new state= 111
I/flutter ( 6145): new state= 16
I/flutter ( 6145): new state= 116
```

So I clearly have 2 separate instances of my StateNotifierProvider, each with it's own state.
I can interact with either instance using `.read(indexProvider.notifier).add(5);`
But I can not use `providerContainer.watch`

Anyhow, do you have a pointer for me on how to get a Consumer widget to monitor `providerContainer`?

**Comment by rrousselGit** [comment]
> In your example code I discovered this: final providerContainer = ProviderContainer();


That's anti pattern.
You're not supposed to create ProviderContainers like that.

**Comment by 3SCadmin** [comment]
Ok,

If I wanted to change a provider's state in `(WidgetsBinding.instance!.addPostFrameCallback((_) {})`
What should I be doing?

**Reply by rrousselGit** [reply]
Use ConsumerStatefulWidget and you should now have a "ref" within your widget

**Reply by 3SCadmin** [reply]
> Use ConsumerStatefulWidget and you should now have a "ref" within your widget

Yes, that works  (I was trying something more creative)

Thinking ahead:  What if I spawn an isolate?
How would my isolate call a provider's inner function to change  a provider's state?

**Reply by rrousselGit** [reply]
Isolates by definition do not share state with the other isolates

As such isolates cannot modify providers

**Reply by 3SCadmin** [reply]
Fair enough, I will try and re-think some of the concepts I came up with.
